pkg/agent/route/route.go

```diff
@@ -0,0 +1,303 @@
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/config"
+	"github.com/vmware-tanzu/antrea/pkg/agent/iptables"
+	"github.com/vmware-tanzu/antrea/pkg/agent/types"
+	"github.com/vmware-tanzu/antrea/pkg/agent/util"
+)
+
+const (
+	// AntreaServiceTable is route table name for Antrea service traffic.
+	AntreaServiceTable = "Antrea-service"
+	// AntreaServiceTableIdx is route table index for Antrea service traffic.
+	AntreaServiceTableIdx = 300
+	routeTableConfigPath  = "/etc/iproute2/rt_tables"
+)
+
+// Client is route client.
+type Client struct {
+	nodeConfig *types.NodeConfig
+}
+
+type serviceRtTableConfig struct {
+	Idx  int
+	Name string
+}
+
+func (s *serviceRtTableConfig) String() string {
+	return fmt.Sprintf("%s: idx %d", s.Name, s.Idx)
+}
+
+func (s *serviceRtTableConfig) IsMainTable() bool {
+	return s.Name == "main"
+}
+
+var (
+	// ServiceRtTable contains Antrea service route table information.
+	ServiceRtTable = &serviceRtTableConfig{Idx: 254, Name: "main"}
+)
+
+// NewClient returns a route client
+func NewClient() *Client {
+	return &Client{}
+}
+
+// Initialize sets up route tables for Antrea.
+func (c *Client) Initialize(nodeConfig *types.NodeConfig) error {
+	c.nodeConfig = nodeConfig
+
+	if config.TrafficEncapMode.SupportsNoEncap() {
+		ServiceRtTable.Idx = AntreaServiceTableIdx
+		ServiceRtTable.Name = AntreaServiceTable
+	}
+
+	if ServiceRtTable.IsMainTable() {
+		_ = c.RemoveServiceRouting()
+		return nil
+	}
+	f, err := os.OpenFile(routeTableConfigPath, os.O_RDWR|os.O_APPEND, 0)
+	if err != nil {
+		klog.Fatalf("Unable to create service route table(open), err %v", err)
+	}
+	defer func() { _ = f.Close() }()
+
+	oldTablesRaw := make([]byte, 1024)
+	bLen, err := f.Read(oldTablesRaw)
+	if err != nil {
+		klog.Fatalf("Unable to create service route table(read), err %v", err)
+	}
+	oldTables := string(oldTablesRaw[:bLen])
+	newTable := fmt.Sprintf("%d %s", ServiceRtTable.Idx, ServiceRtTable.Name)
+
+	if strings.Index(oldTables, newTable) == -1 {
+		if _, err := f.WriteString(newTable); err != nil {
+			klog.Fatalf("Failed to add antrea service route table, err=%v", err)
+		}
+	}
+
+	gwConfig := c.nodeConfig.GatewayConfig
+	if gwConfig != nil && c.nodeConfig.PodCIDR != nil {
+		// add local podCIDR if applicable to service rt table
+		route := &netlink.Route{
+			LinkIndex: util.GetNetLink(gwConfig.Link).Attrs().Index,
+			Scope:     netlink.SCOPE_LINK,
+			Dst:       c.nodeConfig.PodCIDR,
+			Table:     ServiceRtTable.Idx,
+		}
+		if err := netlink.RouteReplace(route); err != nil {
+			klog.Fatalf("Failed to add link route to service table, err=%v", err)
+		}
+	}
+
+	// create ip rule to select route table
+	ipRule := netlink.NewRule()
+	ipRule.IifName = util.GetNetLink(c.nodeConfig.GatewayConfig.Link).Attrs().Name
+	ipRule.Mark = iptables.RtTblSelectorValue
+	ipRule.Table = ServiceRtTable.Idx
+
+	err = netlink.RuleAdd(ipRule)
+	if err != nil {
+		if os.IsExist(err) {
+			klog.Errorf("rule already created")
```

Comment:
added priority value to ip rule to prevent redundant ip rule creation,

Reply:
That doesn't stop duplicate rules from being created:
```
root@dev4docker:~/antrea# ip rule add iif gw0 lookup 310 priority 32756
root@dev4docker:~/antrea# ip rule add iif gw0 lookup 310 priority 32756
root@dev4docker:~/antrea# ip rule
0:      from all lookup local
32756:  from all iif gw0 lookup 310
32756:  from all iif gw0 lookup 310
```
It has to do a check before creating.

In case you miss another comment, the log should NOT be Error if the rule already exists as it could be a normal restart.

Reply:
I think netlink priority maps to pref from ip rule command

root@f1c86e7fe988:/# ip rule add iif eth0 pref 300 table 300
root@f1c86e7fe988:/# ip rule add iif eth0 pref 300 table 300
RTNETLINK answers: File exists
root@f1c86e7fe988:/# echo $?
2

I did test with existing netlink implementation and verified it.

Reply:
I am not sure why ur set up did not detect. netlink does not have api to check for rule existence,

Reply:
thx for the pointer, used netlink.ruleList to detect redundancy.